himportscan/src/HImportScan/ImportScanner.hs

```diff
@@ -69,26 +65,55 @@ instance Aeson.ToJSON ModuleImport where
-   in case scanTokenStream filePath $ lexTokenStream sbuffer loc of
-    Left err -> error err
-    Right ScannedData{moduleName, importedModules, usesTH} ->
-      ScannedImports
-        { filePath = Text.pack filePath
-        , moduleName
-        , importedModules
-        , usesTH
-        }
-
+  if not fileExists
+  then pure Nothing
+  else scanImports filePath =<< Text.readFile filePath
+
+-- TODO[GL]: This function is only in IO because
+-- * we use printBagOfErrors to report an error, but we can easily factor that out
+-- * getImports is in IO, which in turn is only in IO to throw an error
+--   Perhaps we could raise an issue at ghc to make a pure variant.
+scanImports :: FilePath -> Text -> IO (Maybe ScannedImports)
+scanImports filePath contents = do
+  -- TODO[GL]: going through String just because StringBuffer doesn't have a Text interface is not the best
+  -- we could potentially skip that with more effort (e.g. go through ByteString, which is very similar to a StringBuffer)
+  let sb = GHC.stringToStringBuffer $ Text.unpack $ preprocessContents contents
+
+  -- TODO[GL]: Once we're on ghc 9.2 we can get rid of all the things relating to dynFlags, and use the much smaller
+  -- ParserOpts, as getImports no longer depends on DynFlags then.
+  let dynFlags = GHC.Utils.dynFlags
+
+  let
+    -- [GL] The fact that the resulting strings here contain the "-X"s makes me a bit doubtful that this is the right approach,
+    -- but this is what I found for now.
+    usesTH =
+      any (`elem` ["-XTemplateHaskell", "-XQuasiQuotes"]) $
+        map GHC.unLoc $
+          GHC.getOptions dynFlags sb filePath
+  -- TODO: should both of the files here be filePath?
+  GHC.getImports dynFlags sb filePath filePath >>= \case
+    -- It's important that we error in this case, as otherwise the parser has gone terribly wrong probably.
+    -- It's also ok to print the error, as it is usually descriptive and well formatted.
+    -- The way we error here is that we're passing unexpected output to the go library.
+    -- This is far from ideal, however handling this better would require being able to communicate errors better to go.
+    -- TODO[GL]: propagate error instead
+    Left err -> do
+      GHC.printBagOfErrors dynFlags err
+      pure Nothing
+    Right (sourceImports, normalImports, moduleName) -> do
+      pure $ Just ScannedImports
+            { filePath = Text.pack filePath
+            , moduleName = GHC.Utils.moduleNameToText moduleName
+            , importedModules =
+                let
+                  toModuleImport :: (Maybe GHC.FastString, GHC.Located GHC.ModuleName) -> ModuleImport
+                  toModuleImport (mfs, locatedModuleName) =
+                    ModuleImport
+                      (fmap (Text.pack . GHC.unpackFS) mfs)
+                      (GHC.Utils.moduleNameToText locatedModuleName)
+                 in Set.fromList $ map toModuleImport $ sourceImports ++ normalImports
+            , usesTH = usesTH
```

Comment:
This also should work with NameFieldPuns as enabled in this module:
```
- , usesTH = usesTH
+ , usesTH
  }
```

Reply:
Oh yes, I had an expression on the right previously, thanks.